src/Microsoft.AspNetCore.Mvc.Core/Internal/ContentResultExecutor.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    public class ContentResultExecutor
+    {
+        /// <summary>
+        /// The maximum number of characters that are encoded and written to stream in a single pass.
+        /// </summary>
+        public const int MaxCharacterChunkSize = 1024;
+
+        private readonly string DefaultContentType = new MediaTypeHeaderValue("text/plain")
+        {
+            Encoding = Encoding.UTF8
+        }.ToString();
+
+        private readonly ILogger<ContentResultExecutor> _logger;
+
+        public ContentResultExecutor(ILogger<ContentResultExecutor> logger)
+        {
+            _logger = logger;
+        }
+
+        public async Task ExecuteAsync(ActionContext context, ContentResult result)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (result == null)
+            {
+                throw new ArgumentNullException(nameof(result));
+            }
+
+            var response = context.HttpContext.Response;
+
+            string resolvedContentType;
+            Encoding resolvedContentTypeEncoding;
+            ResponseContentTypeHelper.ResolveContentTypeAndEncoding(
```

Comment:
Looks great so far, but I don't like how this helper is implemented. 

1. it always reads the response.ContentType even when result.ContentType is set
1. it always assigns the response.ContentType even if the result.ContentType is not set
1. MediaType.GetEncoding is always called for the default, which is always UTF-8

I'd suggest changing the helper 
```cs
public static Encoding ApplyActionResultContentType(
  HttpResponse response, 
  string actionResultContentType = null, 
  Encoding defaultEncoding = null)
{
  if (!string.IsNullOrEmpty(actionResultContentType))
  {
    response.ContentType = actionResultContentType;
    return 
      MediaType.GetEncoding(actionResultContentType) ?? 
      defaultEncoding ??
      Encoding.UTF8;
  }

  var responseContentType = response.ContentType;
  if (!string.IsNullOrEmpty(responseContentType))   
  {
    return 
      MediaType.GetEncoding(responseContentType) ?? 
      defaultEncoding ?? 
      Encoding.UTF8;
  }

  return defaultEncoding ?? Encoding.UTF8;
}
```